pkg/kubectl/cmd/cmd.go

```diff
@@ -257,7 +262,80 @@ var (
 )
 
 func NewDefaultKubectlCommand() *cobra.Command {
-	return NewKubectlCommand(os.Stdin, os.Stdout, os.Stderr)
+	cmd := NewKubectlCommand(os.Stdin, os.Stdout, os.Stderr)
+
+	if len(os.Args) > 1 {
+		cmdPathPieces := os.Args[1:]
+
+		// traverse path in command tree from args given, attempting
+		// to determine the farthest path that allows extensibility
+		// if dealing with a binary extension.
+		currentCmd := cmd
+		canBeExtended := false
+		for _, target := range cmdPathPieces {
+			subCommandExists := false
+
+			for _, actual := range currentCmd.Commands() {
+				if actual.Name() == target {
+					currentCmd = actual
+					subCommandExists = true
+					break
+				}
+			}
+
+			if !subCommandExists {
+				if canBeExtended {
+					handleEndpointExtensions(cmdPathPieces)
+				}
+				break
+			}
+
+			// if subcommand exists, determine if it allows extensibility
+			if val, ok := currentCmd.Annotations[EndpointCanBeExtendedAnnotation]; ok {
+				canBeExtended = val == "true"
+			} else {
+				canBeExtended = false
+			}
+		}
+	}
+
+	return cmd
+}
+
+func handleEndpointExtensions(cmdArgs []string) {
+	remainingArgs := cmdArgs
+	for idx := range remainingArgs {
+		remainingArgs[idx] = strings.Replace(remainingArgs[idx], "-", "_", -1)
```

Comment:
* should we stop as soon as we encounter an option (an arg starting with `-`)?
* append the normalized form to a new slice, overwriting here modifies cmdArgs and cmdPathPieces and os.Args

Reply:
> Could we instead "move" those from remainingArgs to a separate slice, then append them to the list of args that we ultimately pass to the subprocess? 

we don't have enough info to move flags here. you would also have to move accompanying value args (e.g. move `--foo bar` as a unit, without knowing whether `--foo` is an arg that expects a value).

> I would be in support of allowing flags to be passed as arguments to plugins

absolutely, I'm saying we should not include them in the list of args considered when searching for a named plugin.

If I call `kubectl plugins foo bar --baz qux`, we should start our search with `kubectl-plugins-foo-bar`. If we find `kubectl-plugins-foo`, I'd expect to pass the args `["bar","--baz","qux"]` to it

Reply:
Note that this implies all flags come *after* the command name, which is sometimes-true-and-sometimes-not for kubelet flags, depending on whether they are global flags like --namespace or command-specific flags like --dry-run

Reply:
@liggitt 
> If I call kubectl plugins foo bar --baz qux, we should start our search with kubectl-plugins-foo-bar. If we find kubectl-plugins-foo, I'd expect to pass the args ["bar","--baz","qux"] to it

This was my original approach when searching for a plugin binary, however @deads2k pointed out that if a user has two binaries `kubectl-plugins-foo` and `kubectl-plugins-foo-bar` in his PATH, it is more reasonable for our use-case that we match `kubectl-plugins-foo-bar` over the other one, hence why I switched to reverse matching the binary name.

Also, thinking about flag handling a bit more, another scenario to consider would be when a user specifies a flag [and a potential value for it] in the middle of a command invocation: `kubectl plugin foo --flagname maybeflagvalue bar baz` (binary to match here could be kubectl-plugin-foo-bar)

Reply:
> This was my original approach when searching for a plugin binary, however @deads2k pointed out that if a user has two binaries kubectl-plugins-foo and kubectl-plugins-foo-bar in his PATH, it is more reasonable for our use-case that we match kubectl-plugins-foo-bar over the other one, hence why I switched to reverse matching the binary name.

I agree with the reverse search. In my example, the reverse search would consist of:
* kubectl-plugins-foo-bar
* kubectl-plugins-foo

It would not include `kubectl-plugins-foo-bar-__bar-qux`

Reply:
> Also, thinking about flag handling a bit more, another scenario to consider would be when a user specifies a flag [and a potential value for it] in the middle of a command invocation: kubectl plugin foo --flagname maybeflagvalue bar baz (binary to match here could be kubectl-plugin-foo-bar)

We don't have sufficient information to support that case. That's what I was trying to describe in this comment:

> we don't have enough info to move flags here. you would also have to move accompanying value args (e.g. move --foo bar as a unit, without knowing whether --foo is an arg that expects a value).

Reply:
> We don't have sufficient information to support that case. That's what I was trying to describe in this comment:

Makes sense, will update reverse search to take "flag-value" pairs into account. Thanks